cabal-install/Distribution/Client/ProjectBuilding.hs

```diff
@@ -1166,8 +1168,15 @@ buildInplaceUnpackedPackage verbosity
 
         -- Haddock phase
         whenHaddock $
-          annotateFailureNoLog HaddocksFailed $
-          setup haddockCommand haddockFlags []
+          annotateFailureNoLog HaddocksFailed $ do
+            setup haddockCommand haddockFlags []
+            let haddockTarget = elabHaddockForHackage pkg
+            when (haddockTarget == Cabal.ForHackage) $ do
+              let dest = distDirectory </> name <.> "tar.gz"
```

Comment:
I think it should be `$dist-newstyle/doc/$tarball-name`. Or maybe even `$distBuildDirectory/doc/$tarball-name`.